### PR TITLE
Split fees proportionally in orders between maker and taker

### PR DESCRIPTION
**Pull request by nuevoalex**
I didn't add any new tests since this is covered by the wcl tests pretty well it seems.

### CODE REVIEW

**Comment by joeykrug** [comment]
How's the environment variable thing work
		

	





On Wed, Jul 5, 2017 at 6:13 PM -0700, "Micah Zoltu" <notifications@github.com> wrote:












@MicahZoltu approved this pull request.



Looks correct.  Did you run the WCL fuzzy tests?  You have to export an environment variable for them to run with pytest because they take a long time.  If not, recommend doing so as they may have broken and need updated as well.



In src/trading/takeAskOrder.se:
> @@ -96,11 +96,17 @@ def takeAskOrder(taker: address, orderID: address, market: address, outcome: int
             i += 1
         # sell the complete sets (this will pay fees)
         COMPLETE_SETS.sellCompleteSets(self, market, completeSets)
-        # FIXME: in order to ensure that we can always cover fees, fees should be split proportionately to maker and taker based on how much cash each party will get.
-        # maker gets full share of cash
-        makerShare = safeFxpMul(completeSets, sharePriceLong)
+        # figure out the expected payout before fees
+        cumulativeScale = market.getCompleteSetCostInAttotokens()
+        cost = safeFxpMul(completeSets, cumulativeScale)
+        # get the completeSetFee
+        payout = denominationToken.balanceOf(self)
+        completeSetFee = safeSub(cost, payout)



I wonder if we should add a method to COMPLETE_SETS for getValue(numberOfSets).  I believe there are a couple of places where we need to figure out how much some number of complete sets are worth (cost in this case) and it would be nice if that logic was centralized rather than repeated.  It isn't complex logic, but when reading code it often takes me a second to figure out what it is trying to do.





In tests/trading/test_wcl.py:
> @@ -47,9 +47,11 @@ def test_make_ask_with_shares_take_with_shares(contractsFixture):
     # 3. take ASK order for YES with NO shares
     assert noShareToken.approve(takeOrder.address, fix('1.2'), sender = tester.k2)
     fxpAmountRemaining = takeOrder.publicTakeOrder(askOrderID, ASK, market.address, YES, fix('1.2'), sender = tester.k2)
+    makerFee = fix(completeSetFees, '0.6') / fix('1')
+    takerFee = fix(completeSetFees, '0.4') / fix('1')



Style: Recommend putting this near the top with all of the "setup" code (near where completeSetFees is defined).





In tests/trading/test_wcl.py:
> @@ -47,9 +47,11 @@ def test_make_ask_with_shares_take_with_shares(contractsFixture):
     # 3. take ASK order for YES with NO shares
     assert noShareToken.approve(takeOrder.address, fix('1.2'), sender = tester.k2)
     fxpAmountRemaining = takeOrder.publicTakeOrder(askOrderID, ASK, market.address, YES, fix('1.2'), sender = tester.k2)
+    makerFee = fix(completeSetFees, '0.6') / fix('1')
+    takerFee = fix(completeSetFees, '0.4') / fix('1')



fix(completeSetFees, '0.6') isn't correct.  completeSetFees is already in fixed point, so calling fix on it multiplies it by an extra 10^18 (which you then divide away shortly after).  Instead just do:
makerFee = completeSetFees * 0.6
takerFee = completeSetFees * 0.4



I don't think this will lead to any rounding problems...





In tests/trading/test_wcl.py:
> @@ -198,9 +200,11 @@ def test_make_bid_with_shares_take_with_shares(contractsFixture):
     # 3. take BID order for YES with shares of YES
     assert yesShareToken.approve(takeOrder.address, fix('1.2'), sender = tester.k2)
     leftoverInOrder = takeOrder.publicTakeOrder(orderId, BID, market.address, YES, fix('1.2'), sender = tester.k2)
+    makerFee = fix(completeSetFees, '0.4') / fix('1')
+    takerFee = fix(completeSetFees, '0.6') / fix('1')



Hmm, I see why you don't have this up in the setup now.  Maybe instead name these two variables shortFee and longFee?  Alternatively, leave things as they are.  :)





In tests/trading/test_wcl.py:
> -    (    BID,     NO,  '0.6',  '1.2',  '1.2',    '0',    '0',  '1.2',    '0',  '1.2',    '0',  False,    '0',    '0', '0.48',  False,    '0',    '0','0.70788', lazy_fixture('contractsFixture')),
-    (    BID,     NO,  '0.6',  '1.2',    '0',    '0', '0.72',  '1.2',    '0',    '0', '0.48',  False,    '0',  '1.2',    '0',  False,  '1.2',    '0',      '0', lazy_fixture('contractsFixture')),
-    (    BID,     NO,  '0.6',  '1.2',  '1.2',    '0',    '0',  '1.2',    '0',    '0', '0.48',  False,    '0',    '0', '0.48',  False,  '1.2',    '0',      '0', lazy_fixture('contractsFixture')),
-
-    (    BID,     NO,  '0.6',  '2.4',  '1.2',    '0', '0.72',  '2.4',    '0',  '2.4',    '0',  False,    '0',  '1.2', '0.48',  False,    '0',    '0','1.42788', lazy_fixture('contractsFixture')),
-    (    BID,     NO,  '0.6',  '2.4',  '1.2',    '0', '0.72',  '2.4',    '0',    '0', '0.96',  False,    '0',  '1.2', '0.48',  False,  '2.4',    '0',      '0', lazy_fixture('contractsFixture')),
-    (    BID,     NO,  '0.6',  '2.4',    '0',    '0', '1.44',  '2.4',    '0',  '1.2', '0.48',  False,    '0',  '2.4',    '0',  False,  '1.2',    '0',   '0.72', lazy_fixture('contractsFixture')),
-    (    BID,     NO,  '0.6',  '2.4',  '2.4',    '0',    '0',  '2.4',    '0',  '1.2', '0.48',  False,    '0',    '0', '0.96',  False,  '1.2',    '0','0.70788', lazy_fixture('contractsFixture')),
-
-    (    BID,     NO,  '0.6',  '2.4',  '1.2',    '0', '0.72',  '2.4',    '0',  '1.2', '0.48',  False,    '0',  '1.2', '0.48',  False,  '1.2',    '0','0.70788', lazy_fixture('contractsFixture')),
-
-    (    ASK,    YES,  '0.6',  '1.2',  '1.2',    '0',    '0',  '1.2',    '0',    '0', '0.72',  False,    '0',    '0', '0.72',  False,  '1.2',    '0',      '0', lazy_fixture('contractsFixture')),
-    (    ASK,    YES,  '0.6',  '1.2',    '0',    '0', '0.48',  '1.2',    '0',    '0', '0.72',  False,    '0',  '1.2',    '0',  False,  '1.2',    '0',      '0', lazy_fixture('contractsFixture')),
-    (    ASK,    YES,  '0.6',  '1.2',  '1.2',    '0',    '0',  '1.2',    '0',  '1.2',    '0',  False,    '0',    '0', '0.72',  False,    '0',    '0','0.46788', lazy_fixture('contractsFixture')),
-    (    ASK,    YES,  '0.6',  '1.2',    '0',    '0', '0.48',  '1.2',    '0',  '1.2',    '0',  False,    '0',  '1.2',    '0',  False,    '0',    '0',   '0.48', lazy_fixture('contractsFixture')),
+    # | ------ ORDER ------ |   | ------ MAKER START ------ |   | ------ TAKER START ------ |  | ------ MAKER FINISH -------  |    | ------- TAKER FINISH -------  |



Consider adding an extra hyphen before MAKER FINISH and a space between it and TAKER FINISH so things align marginally better.  :)  Or don't since I need the practice fighting my OCD.

**Comment by MicahZoltu** [comment]
@joeykrug 
```
export INCLUDE_FUZZY_TESTS=1
pytest trading/test_wcl_fuzzy.py
unset INCLUDE_FUZZY_TESTS
```

Or if you want to run all tests (including fuzzy):
```
export INCLUDE_FUZZY_TESTS=1
pytest
unset INCLUDE_FUZZY_TESTS
```

File: src/trading/takeAskOrder.se [hunk 1]
```diff
@@ -96,11 +96,17 @@ def takeAskOrder(taker: address, orderID: address, market: address, outcome: int
...
         COMPLETE_SETS.sellCompleteSets(self, market, completeSets)
-        # FIXME: in order to ensure that we can always cover fees, fees should be split proportionately to maker and taker based on how much cash each party will get.
-        # maker gets full share of cash
-        makerShare = safeFxpMul(completeSets, sharePriceLong)
+        # figure out the expected payout before fees
+        cumulativeScale = market.getCompleteSetCostInAttotokens()
+        cost = safeFxpMul(completeSets, cumulativeScale)
+        # get the completeSetFee
+        payout = denominationToken.balanceOf(self)
+        completeSetFee = safeSub(cost, payout)
```
Comment (nuevoalex) [inline src/trading/takeAskOrder.se]: Seems reasonable to me. I like code that reads easier in general. Will add

File: tests/trading/test_wcl.py [hunk 2]
```diff
@@ -47,9 +47,11 @@ def test_make_ask_with_shares_take_with_shares(contractsFixture):
     # 3. take ASK order for YES with NO shares
     assert noShareToken.approve(takeOrder.address, fix('1.2'), sender = tester.k2)
     fxpAmountRemaining = takeOrder.publicTakeOrder(askOrderID, ASK, market.address, YES, fix('1.2'), sender = tester.k2)
+    makerFee = fix(completeSetFees, '0.6') / fix('1')
+    takerFee = fix(completeSetFees, '0.4') / fix('1')
```
Comment (nuevoalex) [inline tests/trading/test_wcl.py]: Makes sense. Will do.

File: tests/trading/test_wcl.py [hunk 3]
```diff
@@ -198,9 +200,11 @@ def test_make_bid_with_shares_take_with_shares(contractsFixture):
     # 3. take BID order for YES with shares of YES
     assert yesShareToken.approve(takeOrder.address, fix('1.2'), sender = tester.k2)
     leftoverInOrder = takeOrder.publicTakeOrder(orderId, BID, market.address, YES, fix('1.2'), sender = tester.k2)
+    makerFee = fix(completeSetFees, '0.4') / fix('1')
+    takerFee = fix(completeSetFees, '0.6') / fix('1')
```
Comment (nuevoalex) [inline tests/trading/test_wcl.py]: heh, I'll decide something.

File: tests/trading/test_wcl.py [hunk 4]
```diff
@@ -319,36 +323,36 @@ def placeholder_context():
...
-
-    (    BID,     NO,  '0.6',  '2.4',  '1.2',    '0', '0.72',  '2.4',    '0',  '1.2', '0.48',  False,    '0',  '1.2', '0.48',  False,  '1.2',    '0','0.70788', lazy_fixture('contractsFixture')),
-
-    (    ASK,    YES,  '0.6',  '1.2',  '1.2',    '0',    '0',  '1.2',    '0',    '0', '0.72',  False,    '0',    '0', '0.72',  False,  '1.2',    '0',      '0', lazy_fixture('contractsFixture')),
-    (    ASK,    YES,  '0.6',  '1.2',    '0',    '0', '0.48',  '1.2',    '0',    '0', '0.72',  False,    '0',  '1.2',    '0',  False,  '1.2',    '0',      '0', lazy_fixture('contractsFixture')),
-    (    ASK,    YES,  '0.6',  '1.2',  '1.2',    '0',    '0',  '1.2',    '0',  '1.2',    '0',  False,    '0',    '0', '0.72',  False,    '0',    '0','0.46788', lazy_fixture('contractsFixture')),
-    (    ASK,    YES,  '0.6',  '1.2',    '0',    '0', '0.48',  '1.2',    '0',  '1.2',    '0',  False,    '0',  '1.2',    '0',  False,    '0',    '0',   '0.48', lazy_fixture('contractsFixture')),
+    # | ------ ORDER ------ |   | ------ MAKER START ------ |   | ------ TAKER START ------ |  | ------ MAKER FINISH -------  |    | ------- TAKER FINISH -------  |
```
Comment (nuevoalex) [inline tests/trading/test_wcl.py]: But now my OCD is triggered so there is no such option

**Comment by nuevoalex** [comment]
The fuzzy tests do have some failures. I'll fix these.

**Comment by nuevoalex** [comment]
Updated with suggestions addressed and all tests passing

**Pull request merged**